### DISCUSSION TITLE
dependent repeats

### DISCUSSION THREAD

**Discussion by pherk** [Jinntec/Fore#280] [Q&A] [answered]
i have a cal structure (cal >  schedule > agenda > event) and wants to show it in columns.
The dependent repeats have to use index('parent') to access the data subset.
So far so good, the columns for the first items show up.
But the index() does not update if another item in the parent is clicked.
Surely, I'm missing something.

Simplified snippet:
~~~
   <div class="grid">
            <fx-group class="wrapit">
            <h3>Schedules</h3>
              <fx-repeat id="r-sched-id" ref="instance('ical')//cal/schedule">
                <template>
                    <div>
                        <fx-output ref="./global/display/@value"/>
                    </div>
                </template>
              </fx-repeat>
            </fx-group>

            <fx-group class="wrapit">
            <h3>Agendas</h3>
              <fx-repeat id="r-agenda-id" ref="instance('ical')//cal/schedule[index('r-sched-id')]/agenda">
                <template>
                    <div>
                        <fx-output value="substring(./period/start/@value,1,10)"></fx-output>
                        <fx-output value="substring(./period/end/@value,1,10)"></fx-output>
                    </div>
                </template>
              </fx-repeat>
            </fx-group>
....
    </div>
~~~ 

PS: has Fore an 'index-changed' event (as betterforms)? I tried 'path-mutated' but it does not work for that purpose.

regards
Peter

**Comment by JoernT** [comment]
@pherk thanks for the report. It would however be nice if you could provide a complete page as otherwise we need to prepare mockup data first to investigate it. 

We will however look into it.

**Comment by JoernT** [comment]
To be more precise this has to do with the fact that Fore does not track pure UI dependencies - here the repeat refs are dependant but that isn't picked up by the Dependency Graph with handles only pure data dependencies established via fx-binds.

You can however update the other repeat with explicitly refreshing the other repeat - we'll prepare the right snippet to make that work....

**Reply by pherk** [reply]
your explanation helps a lot.
Is this in repeat > template > action also canonical?
~~~
                <fx-action event="item-changed">
                    <fx-message>index changed</fx-message>
                    <fx-update/>
                    <fx-refresh/>
                </fx-action>
~~~
or should I use fx-bind as you suggested? 
Do you expect significant difference in performance?
I promise that my next question will have a complete test page ;-)

**Reply by JoernT** [reply]
Yes, this would be a solution. It causes a full refresh but shouldn't cause dramatic performance penalties unless the repeat contains hundreds of entries which i don't expect from your use case.

A bind wouldn't help much here i guess as the second repeat will have to fully refresh anyway when the first one changes.

Potentially you could speed up the above by using `<fx-refresh control="r-agenda-id">` which will only refresh the second repeat. But if your UI mainly consists of the repeat there wouldn't be a noticable difference.

**Comment by DrRataplan** [comment]
I think the main issue here is that the index function specifically has no way to be reran. I'd say that because of this, an index change in a repeat should _always_ cause a full update.

Annoying, but I'd choose the performance hit over these strange bugs

**Reply by DrRataplan** [reply]
I made a PR to do this in #281

**Comment by pherk** [comment, marked as answer]
thank you both for your insights /inputs.
I have grep'ed after the specific events for fx-repeat in source and saw their params (btw not yet in Fore Documentation - Events#Repeat Events ).
Luckily I now have three solutions..

**Reply by JoernT** [reply]
thanks for pointing that out. worked a bit on docs
